contracts/SponsorFaucet.sol

```diff
@@ -131,95 +97,80 @@ contract SponsorFaucet is Ownable, Pausable, ReentrancyGuard, WhitelistedRole {
-        general_gas_bound = gasBound;
-        general_collateral_bound = collateralBound;
-        general_upper_bound = upperBound;
+        require(largeBounds.upper_bound.mul(1000) <= largeBounds.gas_bound, "upperBound too high");
+        dapp_bounds[large] = largeBounds;
     }
 
-    //set bounds for special contracts
-    function setSpecialBounds(
+    //set bounds for custom contracts
+    function setCustomBounds(
         address addr,
-        uint256 gasTotalLimit,
-        uint256 collateralTotalLimit,
-        uint256 gasBound,
-        uint256 collateralBound,
-        uint256 upperBound
+        bounds memory customBounds
     ) public {
         _onlyWhitelistAdmin();
         //rule by internal contract
-        require(upperBound.mul(1000) <= gasBound, "upperBound too high");
-        special_dapps[addr].gas_total_limit = gasTotalLimit;
-        special_dapps[addr].collateral_total_limit = collateralTotalLimit;
-        special_dapps[addr].gas_bound = gasBound;
-        special_dapps[addr].collateral_bound = collateralBound;
-        special_dapps[addr].upper_bound = upperBound;
+        require(customBounds.upper_bound.mul(1000) <= customBounds.gas_bound, "upperBound too high");
+        dapp_bounds[addr] = customBounds;
     }
 
 
     /* ===== Public utility functions ===== */
-    function addGeneralContracts(address[] memory addAddr) public {
+    function addLargeContracts(address[] memory addAddr) public {
         _onlyWhitelistAdmin();
         for(uint256 i = 0; i < addAddr.length; i++) {
-            general_contracts.push(addAddr[i]);
+            large_contracts[addAddr[i]] = true;
+            if(custom_contracts[addAddr[i]]) {
+                delete dapp_bounds[addAddr[i]];
+            }
+            custom_contracts[addAddr[i]] = false;
         }
     }
 
-    function addSpecialContracts(address[] memory addAddr) public {
+    function addCustomContracts(address[] memory addAddr, bounds[] memory addrBounds) public {
         _onlyWhitelistAdmin();
+        require(addAddr.length == addrBounds.length, "length not match");
         for(uint256 i = 0; i < addAddr.length; i++) {
-            special_contracts.push(addAddr[i]);
+            custom_contracts[addAddr[i]] = true;
+            large_contracts[addAddr[i]] = false;
+            dapp_bounds[addAddr[i]] = addrBounds[i];
         }
     }
 
-    function specialContractsCount() public view returns (uint256) {
-        return special_contracts.length;
+    function removeLargeContract(address[] memory removeAddr) public {
+        _onlyWhitelistAdmin();
+        for(uint256 i = 0; i < removeAddr.length; i++) {
+            delete large_contracts[removeAddr[i]];
+            delete dapp_bounds[removeAddr[i]];
+        }
     }
 
-    function generalContractsCount() public view returns (uint256) {
-        return general_contracts.length;
+    function removeCustomContract(address[] memory removeAddr) public {
```

Comment:
Why not use a single function to remove `large` and `custom` sponsorship?

Reply:
To rm **large** only large_contracts changed, while to rm **custom** a **bounds** needs to be deleted.

Reply:
For large sponsorship case, the address is 0x0000...0001. It's ok to provide a single function for removal purpose, just as what you designed for setBounds.